### ISSUE TITLE
Trying to create more OSDs than asked for

### ISSUE THREAD

**Issue by cite** [rook/rook#10665]
**Is this a bug report or feature request?**
* Bug Report

**Deviation from expected behavior:**

In a host based cluster with OSDs located on LVMs, after the cluster was successfully built, a reconciliation run is trying to add even more OSDs, and failing:

```
2022-08-02 09:49:35.891930 E | op-osd: failed to provision OSD(s) on node fra1pro-k8s-worker-6.k8s.fra1pro.financial.com. &{OSDs:[] Status:failed PvcBackedOSD:false Message:failed to configure devices: failed to initialize osd: failed to run ceph-volume raw command. Running command: /usr/bin/ceph-authtool --gen-print-key
Running command: /usr/bin/ceph --cluster ceph --name client.bootstrap-osd --keyring /var/lib/ceph/bootstrap-osd/ceph.keyring -i - osd new c3c6e3a2-a838-4910-a6b2-6075ce841c7e
Running command: /usr/bin/ceph-authtool --gen-print-key
Running command: /usr/bin/mount -t tmpfs tmpfs /var/lib/ceph/osd/ceph-6
Running command: /usr/bin/chown -R ceph:ceph /dev/dm-5
Running command: /usr/bin/ln -s /dev/dm-5 /var/lib/ceph/osd/ceph-6/block
Running command: /usr/bin/ceph --cluster ceph --name client.bootstrap-osd --keyring /var/lib/ceph/bootstrap-osd/ceph.keyring mon getmap -o /var/lib/ceph/osd/ceph-6/activate.monmap
 stderr: got monmap epoch 5
Running command: /usr/bin/ceph-authtool /var/lib/ceph/osd/ceph-6/keyring --create-keyring --name osd.6 --add-key AQAs8+hi7vErNxAAEXzVZx/Kx0S5Uq+6uv4+og==
 stdout: creating /var/lib/ceph/osd/ceph-6/keyring
added entity osd.6 auth(key=AQAs8+hi7vErNxAAEXzVZx/Kx0S5Uq+6uv4+og==)
Running command: /usr/bin/chown -R ceph:ceph /var/lib/ceph/osd/ceph-6/keyring
Running command: /usr/bin/chown -R ceph:ceph /var/lib/ceph/osd/ceph-6/
Running command: /usr/bin/ceph-osd --cluster ceph --osd-objectstore bluestore --mkfs -i 6 --monmap /var/lib/ceph/osd/ceph-6/activate.monmap --keyfile - --osd-data /var/lib/ceph/osd/ceph-6/ --osd-uuid c3c6e3a2-a838-4910-a6b2-6075ce841c7e --setuser ceph --setgroup ceph
 stderr: 2022-08-02T09:49:34.249+0000 7f3c95eb1200 -1 bluestore(/var/lib/ceph/osd/ceph-6/) _open_db_and_around failed to load os-type: (2) No such file or directory
 stderr: 2022-08-02T09:49:34.249+0000 7f3c95eb1200 -1 bluestore(/var/lib/ceph/osd/ceph-6/) mkfs fsck found fatal error: (2) No such file or directory
 stderr: 2022-08-02T09:49:34.249+0000 7f3c95eb1200 -1 OSD::mkfs: ObjectStore::mkfs failed with error (2) No such file or directory
 stderr: 2022-08-02T09:49:34.250+0000 7f3c95eb1200 -1  ** ERROR: error creating empty object store in /var/lib/ceph/osd/ceph-6/: (2) No such file or directory
--> Was unable to complete a new OSD, will rollback changes
Running command: /usr/bin/ceph --cluster ceph --name client.bootstrap-osd --keyring /var/lib/ceph/bootstrap-osd/ceph.keyring osd purge-new osd.6 --yes-i-really-mean-it
 stderr: purged osd.6
Traceback (most recent call last):
  File "/usr/sbin/ceph-volume", line 11, in <module>
    load_entry_point('ceph-volume==1.0.0', 'console_scripts', 'ceph-volume')()
  File "/usr/lib/python3.6/site-packages/ceph_volume/main.py", line 40, in __init__
    self.main(self.argv)
  File "/usr/lib/python3.6/site-packages/ceph_volume/decorators.py", line 59, in newfunc
    return f(*a, **kw)
  File "/usr/lib/python3.6/site-packages/ceph_volume/main.py", line 152, in main
    terminal.dispatch(self.mapper, subcommand_args)
  File "/usr/lib/python3.6/site-packages/ceph_volume/terminal.py", line 194, in dispatch
    instance.main()
  File "/usr/lib/python3.6/site-packages/ceph_volume/devices/raw/main.py", line 32, in main
    terminal.dispatch(self.mapper, self.argv)
  File "/usr/lib/python3.6/site-packages/ceph_volume/terminal.py", line 194, in dispatch
    instance.main()
  File "/usr/lib/python3.6/site-packages/ceph_volume/devices/raw/prepare.py", line 169, in main
    self.safe_prepare(self.args)
  File "/usr/lib/python3.6/site-packages/ceph_volume/devices/raw/prepare.py", line 91, in safe_prepare
    self.prepare()
  File "/usr/lib/python3.6/site-packages/ceph_volume/decorators.py", line 16, in is_root
    return func(*a, **kw)
  File "/usr/lib/python3.6/site-packages/ceph_volume/devices/raw/prepare.py", line 134, in prepare
    tmpfs,
  File "/usr/lib/python3.6/site-packages/ceph_volume/devices/raw/prepare.py", line 68, in prepare_bluestore
    db=db
  File "/usr/lib/python3.6/site-packages/ceph_volume/util/prepare.py", line 481, in osd_mkfs_bluestore
    raise RuntimeError('Command failed with exit code %s: %s' % (returncode, ' '.join(command)))
RuntimeError: Command failed with exit code 250: /usr/bin/ceph-osd --cluster ceph --osd-objectstore bluestore --mkfs -i 6 --monmap /var/lib/ceph/osd/ceph-6/activate.monmap --keyfile - --osd-data /var/lib/ceph/osd/ceph-6/ --osd-uuid c3c6e3a2-a838-4910-a6b2-6075ce841c7e --setuser ceph --setgroup ceph: exit status 1}
2022-08-02 09:49:35.891958 I | op-k8sutil: format and nodeName longer than 63 chars, nodeName fra1pro-k8s-worker-6.k8s.fra1pro.financial.com will be 18a40a276db7618c5ffd6fd2d8ca0b8c
```

However, there are already 5 valid OSDs:

```
[rook@rook-ceph-tools-69c54d54bc-rz6zs /]$ ceph osd tree
ID   CLASS  WEIGHT    TYPE NAME                                                     STATUS  REWEIGHT  PRI-AFF
 -1         15.00000  root default                                                                           
 -8          3.00000      rack pa18                                                                          
 -7          3.00000          host fra1pro-k8s-worker-6-k8s-fra1pro-financial-com                            
  1    hdd   3.00000              osd.1                                                 up   1.00000  1.00000
 -4          6.00000      rack pa19                                                                          
-19          3.00000          host fra1pro-k8s-worker-8-k8s-fra1pro-financial-com                            
  3    hdd   3.00000              osd.3                                                 up   1.00000  1.00000
 -3          3.00000          host fra1pro-k8s-worker-9-k8s-fra1pro-financial-com                            
  4    hdd   3.00000              osd.4                                                 up   1.00000  1.00000
-12          3.00000      rack pa20                                                                          
-11          3.00000          host fra1pro-k8s-worker-10-k8s-fra1pro-financial-com                           
  0    hdd   3.00000              osd.0                                                 up   1.00000  1.00000
-16          3.00000      rack pa25                                                                          
-15          3.00000          host fra1pro-k8s-worker-12-k8s-fra1pro-financial-com                           
  2    hdd   3.00000              osd.2                                                 up   1.00000  1.00000
```

And we also didn't ask for more than 5 OSDs:

```
  storage:
    nodes:
    - devices:
      - name: /dev/mapper/vg_root-lv_k8s--rook--osd
      name: fra1pro-k8s-worker-6.k8s.fra1pro.financial.com
      resources: {}
    - devices:
      - name: /dev/mapper/vg_root-lv_k8s--rook--osd
      name: fra1pro-k8s-worker-8.k8s.fra1pro.financial.com
      resources: {}
    - devices:
      - name: /dev/mapper/vg_root-lv_k8s--rook--osd
      name: fra1pro-k8s-worker-9.k8s.fra1pro.financial.com
      resources: {}
    - devices:
      - name: /dev/mapper/vg_root-lv_k8s--rook--osd
      name: fra1pro-k8s-worker-10.k8s.fra1pro.financial.com
      resources: {}
    - devices:
      - name: /dev/mapper/vg_root-lv_k8s--rook--osd
      name: fra1pro-k8s-worker-12.k8s.fra1pro.financial.com
      resources: {}
    useAllDevices: false
```
(note: the helm chart's values.yaml (see attachment had `useAllNodes: false`, and I just now set this on the cluster CRD, but that was after the failed reconciliation run, and after the value wasn't added by helm)

**Expected behavior:**

Only the specified number of nodes is created.

**How to reproduce it (minimal and precise):**
Happened with two different 1.23.7 clusters, both on Rook 1.9.8.

**Environment**:
* OS (e.g. from /etc/os-release): Oracle Enterprise Linux 8.6
* Kernel (e.g. `uname -a`): `Linux fra1pro-k8s-worker-6 5.15.0-0.30.20.el8uek.x86_64 #2 SMP Sun Jul 10 12:13:30 PDT 2022 x86_64 x86_64 x86_64 GNU/Linux`
* Cloud provider or hardware configuration: HPE Proliatn DL360 Gen10
* Rook version (use `rook version` inside of a Rook Pod): rook: v1.9.8/go: go1.17.12
* Storage backend version (e.g. for ceph do `ceph -v`): ceph version 16.2.10 (45fa1a083152e41a408d15505f594ec5f1b4fe17) pacific (stable)
* Kubernetes version (use `kubectl version`): 1.23.7
* Kubernetes cluster type (e.g. Tectonic, GKE, OpenShift): Rancher on bare metal
* Storage backend status (e.g. for Ceph use `ceph health` in the [Rook Ceph toolbox]: HEALTH_OK

[operator.log](https://github.com/rook/rook/files/9241767/operator.log)
[cluster-yaml.txt](https://github.com/rook/rook/files/9241772/cluster-yaml.txt)
[ceph-cluster-values-fra1pro-yaml.txt](https://github.com/rook/rook/files/9241783/ceph-cluster-values-fra1pro-yaml.txt) (that's the helm chart's values.yaml file)

**Comment by cite** [comment]
Here are the logs for the second cluster.
[operator-chi1pro.log](https://github.com/rook/rook/files/9241800/operator-chi1pro.log)
[cluster-chi1pro-crd-yaml.txt](https://github.com/rook/rook/files/9241802/cluster-chi1pro-crd-yaml.txt)
[ceph-cluster-values-chi1pro-helm.txt](https://github.com/rook/rook/files/9241803/ceph-cluster-values-chi1pro-helm.txt)

**Comment by cite** [comment]
```
[2022-08-01 12:05:26,467][ceph_volume.api.lvm][WARNING] device is not part of ceph: </dev/vg_root/lv_k8s-rook-osd>
```

**Comment by cite** [comment]
Logs from OSD prepare pod.
[rook-ceph-osd-prepare-94a5152b5c9ac0dfac79dc125ca0f3e4-x62n4.log](https://github.com/rook/rook/files/9242044/rook-ceph-osd-prepare-94a5152b5c9ac0dfac79dc125ca0f3e4-x62n4.log)

**Comment by cite** [comment]
Interestingly, the volumes are found from within an OSD:

```
[root@rook-ceph-osd-0-8479d47d4b-7l2kj ceph]# ceph-volume raw list --format json
{
    "e71e58bb-26ed-4303-b4f4-16bec8cc6628": {
        "ceph_fsid": "aac8868b-2a4f-42d5-8856-4da5622695e1",
        "device": "/dev/mapper/vg_root-lv_k8s--rook--osd",
        "osd_id": 0,
        "osd_uuid": "e71e58bb-26ed-4303-b4f4-16bec8cc6628",
        "type": "bluestore"
    }
}
[root@rook-ceph-osd-0-8479d47d4b-7l2kj ceph]# ceph-volume lvm list --format json
{}
```

**Comment by satoru-takeuchi** [comment]
@cite I'm trying to reproduce this problem. Please wait for a while.

**Comment by satoru-takeuchi** [comment]
I tried to create an OSD on LVM in a host-based cluster and restart oeprators (and prepare pods) several times but it worked fine. Hmm...

I'll try to again with the environment as same as yours.

> Interestingly, the volumes are found from within an OSD:

OSD on LVM is so-called "raw mode" OSD. So it is listed in `ceph-volume raw list` which is for raw mode osds. However, it's not listed in `ceph-volume lvm list` which is for lvm mode OSDs.

**Comment by cite** [comment]
I believe the problem might be that the OSD prepare job is only looking for Ceph OSDs in the output of `ceph-volume lvm list`, where the created volumes are not listed. It then seems to proceed directly to trying to create a new OSD, which, of course, fails. At least that's my reading from the OSD prepare job's output.

**Comment by satoru-takeuchi** [comment]
I see. `ceph volume raw list` is not called in the log of your osd prepare pod. In my environment, this command is called in od prepare pod as expected.

```
...
2022-08-09 08:20:17.532741 I | cephosd: no new devices to configure. returning devices already configured with ceph-volume.
2022-08-09 08:20:17.533101 D | exec: Running command: stdbuf -oL ceph-volume --log-path /tmp/ceph-log lvm list  --format json
2022-08-09 08:20:17.864927 D | cephosd: {}
2022-08-09 08:20:17.864968 I | cephosd: 0 ceph-volume lvm osd devices configured on this node
2022-08-09 08:20:17.864986 D | exec: Running command: stdbuf -oL ceph-volume --log-path /tmp/ceph-log raw list --format json
2022-08-09 08:20:18.395455 D | cephosd: {
    "1f194232-740c-4889-a3c7-05259c178afa": {
        "ceph_fsid": "cf9abfa6-2591-4bf5-9d4c-3616f2c70a45",
        "device": "/dev/mapper/testvg-testlv",
        "osd_id": 0,
        "osd_uuid": "1f194232-740c-4889-a3c7-05259c178afa",
        "type": "bluestore"
    }
}
2022-08-09 08:20:18.395573 D | exec: Running command: lsblk /dev/mapper/testvg-testlv --bytes --nodeps --pairs --paths --output SIZE,ROTA,RO,TYPE,PKNAME,NAME,KNAME,MOUNTPOINT,FSTYPE
...
```

**Comment by cite** [comment]
It might already be too late at that point, since it previously is declared available already:

```
2022-08-02 12:09:26.337041 D | exec: Running command: lsblk /dev/mapper/vg_root-lv_k8s--rook--osd --bytes --nodeps --pairs --paths --output SIZE,ROTA,RO,TYPE,PKNAME,NAME,KNAME,MOUNTPOINT,FSTYPE
2022-08-02 12:09:26.342329 D | sys: lsblk output: "SIZE=\"2199023255552\" ROTA=\"1\" RO=\"0\" TYPE=\"lvm\" PKNAME=\"\" NAME=\"/dev/mapper/vg_root-lv_k8s--rook--osd\" KNAME=\"/dev/dm-5\" MOUNTPOINT=\"\" FSTYPE=\"\""
2022-08-02 12:09:26.342350 D | exec: Running command: dmsetup info -c --noheadings -o name /dev/mapper/vg_root-lv_k8s--rook--osd
2022-08-02 12:09:26.343858 D | exec: Running command: dmsetup splitname --noheadings vg_root-lv_k8s--rook--osd
2022-08-02 12:09:26.345350 D | exec: Running command: ceph-volume lvm list --format json vg_root/lv_k8s-rook-osd
2022-08-02 12:09:26.987267 I | cephosd: device "dm-5" is available.
2022-08-02 12:09:26.987298 I | cephosd: "/dev/mapper/vg_root-lv_k8s--rook--osd" found in the desired devices (matched by link: "/dev/mapper/vg_root-lv_k8s--rook--osd")
2022-08-02 12:09:26.987305 I | cephosd: device "dm-5" is selected by the device filter/name "/dev/mapper/vg_root-lv_k8s--rook--osd"
```

**Comment by satoru-takeuchi** [comment]
Probably I found the root cause.

The output of `lsblk` doesn't show the expected FSTYPE "ceph_bluestore" for "/dev/dm-5", on which OSD exist,

```
2022-08-02 12:09:26.250726 D | exec: Running command: lsblk /dev/dm-5 --bytes --nodeps --pairs --paths --output SIZE,ROTA,RO,TYPE,PKNAME,NAME,KNAME,MOUNTPOINT,FSTYPE
2022-08-02 12:09:26.255804 D | sys: lsblk output: "SIZE=\"2199023255552\" ROTA=\"1\" RO=\"0\" TYPE=\"lvm\" PKNAME=\"\" NAME=\"/dev/mapper/vg_root-lv_k8s--rook--osd\" KNAME=\"/dev/dm-5\" MOUNTPOINT=\"\" FSTYPE=\"\""
```

In my environment (the device name is "/dev/dm-0" here):
```
2022-08-10 08:09:27.212390 D | exec: Running command: lsblk /dev/dm-0 --bytes --nodeps --pairs --paths --output SIZE,ROTA,RO,TYPE,PKNAME,NAME,KNAME,MOUNTPOINT,FSTYPE
2022-08-10 08:09:27.216260 D | sys: lsblk output: "SIZE=\"12880707584\" ROTA=\"1\" RO=\"0\" TYPE=\"lvm\" PKNAME=\"\" NAME=\"/dev/mapper/testvg-test--lv\" KNAME=\"/dev/dm-0\" MOUNTPOINT=\"\" FSTYPE=\"ceph_bluestore\""
```

I guess that it's because your "/dev/dm-5" is recognized as an ATARI partition (see "ID_PART_TABLE_TYPE=atari" in the following command output).

```
2022-08-02 12:09:26.255819 D | exec: Running command: udevadm info --query=property /dev/dm-5
2022-08-02 12:09:26.259347 D | sys: udevadm info output: "DEVLINKS=/dev/mapper/vg_root-lv_k8s--rook--osd /dev/disk/by-id/dm-name-vg_root-lv_k8s--rook--osd /dev/vg_root/lv_k8s-rook-osd /dev/disk/by-id/dm-uuid-LVM-s6h63yEWYvSbG1RhEWk7UW7NBqbZNJR5LM00i74Cv6yy02KyYDkNpWlkr9mDLSfN\nDEVNAME=/dev/dm-5\nDEVPATH=/devices/virtual/block/dm-5\nDEVTYPE=disk\nDISKSEQ=9\nDM_LV_NAME=lv_k8s-rook-osd\nDM_NAME=vg_root-lv_k8s--rook--osd\nDM_SUSPENDED=0\nDM_UDEV_DISABLE_LIBRARY_FALLBACK_FLAG=1\nDM_UDEV_PRIMARY_SOURCE_FLAG=1\nDM_UDEV_RULES_VSN=2\nDM_UUID=LVM-s6h63yEWYvSbG1RhEWk7UW7NBqbZNJR5LM00i74Cv6yy02KyYDkNpWlkr9mDLSfN\nDM_VG_NAME=vg_root\nID_PART_TABLE_TYPE=atari\nMAJOR=252\nMINOR=5\nSUBSYSTEM=block\nTAGS=:systemd:\nUSEC_INITIALIZED=13595457"
```

I'll try to bypass this problem later.

**Comment by cite** [comment]
Ah, interesting find. According to `blkid -k`, the `ceph_bluestore` type is not supported. OEL 8.6 ships with `blkid from util-linux 2.32.1  (libblkid 2.32.1, 16-Jul-2018)`.

**Comment by satoru-takeuchi** [comment]
OK, So is it possible to use the newer version of `util-linux`? Then this problem would disappear.

**Comment by cite** [comment]
I don't believe it is. The package seems to be integrated pretty deep into the distribution, and I didn't find any backports for it.

**Comment by satoru-takeuchi** [comment]
@cite Could you try this rook-operator container ("satorutakeuchi/ceph:v1.9.8-fix-10665") as a hotfix?

https://hub.docker.com/layers/269763522/satorutakeuchi/ceph/v1.9.8-fix-10665/images/sha256-2c6bf8d5aaf3a37bb8a4d213bd441f2532d22f3a45b55ba06846f9d804a78cf6?context=repo

This is rook v1.9.8 plus the following patch to add an extra check for the existence of an OSD.
https://github.com/rook/rook/compare/v1.9.8...satoru-takeuchi:rook:v1.9.8-fix-10665

**Comment by satoru-takeuchi** [comment]
OK, I'll make an official fix later. Let's keep this ticket open till then.

**Comment by satoru-takeuchi** [comment]
Probably my co-worker and I will create the official fix.

**Comment by satoru-takeuchi** [comment]
@cite The fix for this problem was merged into master and was backported to both v1.9 and v1.10 release branch. Please update your hotfix container to the next official release (probably v1.9.11 and v1.10.0 or later).
